src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/MemberInfoExtensions.cs

```diff
@@ -67,17 +67,28 @@ public static bool IsDictionaryValueNonNullable(this MemberInfo memberInfo)
             if (memberType.IsValueType) return false;
 
             var nullableAttribute = memberInfo.GetNullableAttribute();
+            var valueArgument = memberType.GetGenericArguments()[1];
+            var valueArgumentIsNullable = valueArgument.IsGenericType && valueArgument.GetGenericTypeDefinition() == typeof(Nullable<>);
 
             if (nullableAttribute == null)
             {
-                return memberInfo.GetNullableFallbackValue();
+                return !valueArgumentIsNullable && memberInfo.GetNullableFallbackValue();
             }
 
-            if (nullableAttribute.GetType().GetField(NullableFlagsFieldName) is FieldInfo field &&
-                field.GetValue(nullableAttribute) is byte[] flags &&
-                flags.Length == 3 && flags[2] == 1)
+            if (nullableAttribute.GetType().GetField(NullableFlagsFieldName) is FieldInfo field)
             {
-                return true;
+                if (field.GetValue(nullableAttribute) is byte[] flags)
+                {
+                    // Ref.: https://github.com/dotnet/roslyn/blob/main/docs/features/nullable-metadata.md
+                    if (flags.Length == 2)  // Value in the dictionary is a value type.
+                    {
+                        return !valueArgumentIsNullable;
+                    }
```

Comment:
I agree that it is confusing. My observations in the debugger show that there are only `3` items in the array if all `3` items are _reference_ types. So the `Dictionary<string, object>` example on the line you linked in spec will have `3` items (one for the `Dictionary` itself, one for the `string` key and one for the `object` value since all `3` are reference types). However if the value in the `Dictionary` is a value type (e.g. `Dictionary<string, int>`), then the observed arity of the array is only `2` which seems to align with [this](https://github.com/dotnet/roslyn/blob/af7b0ebe2b0ed5c335a928626c25620566372dd1/docs/features/nullable-metadata.md?plain=1#L45) talking about them being "skipped". Although - somewhat confusingly - if the value in the `Dictionary` is a nullable value type (e.g. `Dictionary<string, int?>`) then that is still skipped (the arity is still `2`) despite the type of the value being a `Nullable<T>` - i.e. a generic value type! So - in summary, I agree with the confusion, but my observations in the debugger show that we get arity `2` when the Dictionary's value is a value type (or nullable value) type and the test collateral I added exercises that code branch.

Curious to hear your thoughts on this!

Reply:
If that's the way it works then that's fine!

Could you distill this summary into a comment in the code though please? Then in the future someone doesn't have to dig through the spec to try and understand what's happening. Maybe use some local named constants for the magic numbers too?

Reply:
Yep, sounds good. It's funny, I did an initial iteration that I didn't push that was a little more verbose, but worried I was writing too much prose! I'll do a refactor

Reply:
Pushed a refactor with comments and magic numbers removed.